compiler/rustc_middle/src/mir/interpret/allocation.rs

```diff
@@ -825,12 +692,400 @@ impl InitMask {
         self.len += amount;
         self.set_range_inbounds(start, start + amount, new_state); // `Size` operation
     }
+
+    /// Returns the index of the first bit in `start..end` (end-exclusive) that is equal to is_init.
+    fn find_bit(&self, start: Size, end: Size, is_init: bool) -> Option<Size> {
+        /// A fast implementation of `find_bit`,
+        /// which skips over an entire block at a time if it's all 0s (resp. 1s),
+        /// and finds the first 1 (resp. 0) bit inside a block using `trailing_zeros` instead of a loop.
+        ///
+        /// Note that all examples below are written with 8 (instead of 64) bit blocks for simplicity,
+        /// and with the least significant bit (and lowest block) first:
+        ///
+        ///          00000000|00000000
+        ///          ^      ^ ^      ^
+        ///   index: 0      7 8      15
+        ///
+        /// Also, if not stated, assume that `is_init = true`, that is, we are searching for the first 1 bit.
+        fn find_bit_fast(
+            init_mask: &InitMask,
+            start: Size,
+            end: Size,
+            is_init: bool,
+        ) -> Option<Size> {
+            /// Search one block, returning the index of the first bit equal to `is_init`.
+            fn search_block(
+                bits: Block,
+                block: usize,
+                start_bit: usize,
+                is_init: bool,
+            ) -> Option<Size> {
+                // For the following examples, assume this function was called with:
+                //   bits = 11011100
+                //   start_bit = 3
+                //   is_init = false
+                // Note again that the least significant bit is written first,
+                // which is backwards compared to how we normally write numbers.
```

Comment:
Why this choice? It also makes `trailing_zeros` somewhat confusing since in this representation it's actually about leading zeros.^^

Reply:
Yeah this really pained me to write...

I think for the other diagrams like:
```
       00000000|00000000
       ^      ^ ^      ^
index: 0      7 8      15
```
it would be more confusing with the MSB first, because then I'd have to write the blocks in reverse, to have the ranges be contiguous:
```
       00000000|00000000
       ^      ^ ^      ^
index: 15     8 7      0
         ^~~~~~~~~~~~^
        end        start
```

And then I made `search_block`'s comments use the same order to be consistent.

But I guess, since I have a disclaimer here anyways, I can just say "Note that, for the examples in this function, the most significant bit is written first, which is backwards compared to the other comments in `find_bit`"

Reply:
So the real problem here is that the it indexing starts at the LSB... I feel like after a preface explaining that, at least around code that does `trailing_zeros` or `<<`, it'd be better to write bits in the order in which "Rust thinks of them".